group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilder.java

```diff
@@ -0,0 +1,390 @@
+
+            // Only re-add the member to the end of the queue if it's still available for assignment.
+            for (int i = 0; i < unfilledMembers.size() && !assigned; i++) {
+                String memberId = roundRobinMembers.poll();
+                if (unfilledMembers.get(memberId) > 0) {
+                    assignPartitionToMember(memberId, partition);
+                    assigned = true;
+                }
+                if (unfilledMembers.containsKey(memberId)) {
+                    roundRobinMembers.add(memberId);
+                }
+            }
+        });
+    }
+
+    /**
+     * Assigns the specified partition to the given member.
+     *
+     * <p>
+     * If the member has met their allocation quota, the member is removed from the
+     * tracking map of members with their remaining allocations.
+     * Otherwise, the count of remaining partitions that can be assigned to the member is updated.
+     * </p>
+     *
+     * @param memberId      The Id of the member to which the partition will be assigned.
+     * @param partition     The partition to be assigned.
+     */
+    private void assignPartitionToMember(String memberId, TopicIdPartition partition) {
+        newAssignment.get(memberId)
+            .targetPartitions()
+            .computeIfAbsent(partition.topicId(), __ -> new HashSet<>())
+            .add(partition.partition());
+
+        int remainingPartitionCount = unfilledMembers.get(memberId) - 1;
+        if (remainingPartitionCount == 0) {
+            unfilledMembers.remove(memberId);
+        } else {
+            unfilledMembers.put(memberId, remainingPartitionCount);
+        }
+    }
+
+    /**
+     * Determines which members can still be assigned partitions to meet the full quota.
+     *
+     * @return A map of member IDs and their capacity for additional partitions.
+     */
+    private Map<String, Integer> computeUnfilledMembers() {
+        Map<String, Integer> unfilledMembers = new HashMap<>();
+
+        potentiallyUnfilledMembers.forEach((memberId, remaining) -> {
+            if (remainingMembersToGetAnExtraPartition > 0) {
+                remaining++;
+                remainingMembersToGetAnExtraPartition--;
+            }
+            if (remaining > 0) {
+                unfilledMembers.put(memberId, remaining);
+            }
+        });
+
+        return unfilledMembers;
+    }
+
+    /**
+     * This method compares the full list of partitions against the set of already
+     * assigned sticky partitions to identify those that still need to be allocated.
+     *
+     * @param allAssignedStickyPartitions   Set of partitions that have already been assigned.
+     * @return List of unassigned partitions.
+     */
+    private List<TopicIdPartition> computeUnassignedPartitions(Set<TopicIdPartition> allAssignedStickyPartitions) {
+        List<TopicIdPartition> unassignedPartitions = new ArrayList<>();
+        List<Uuid> sortedAllTopics = new ArrayList<>(subscriptionList);
```

Comment:
personally, the word "all" does not add much value to variable names. how's `sortedTopics`

also, what's the benefit of sorting the topics?

Reply:
all means every single topic right, if it was just sorted topics, which/whose topics?

Reply:
It makes sure that all the unassigned partitions are present topic wise and then when we do a rr distribution while assigning these partitions they are distributed more evenly by topic as well

Reply:
> all means every single topic right, if it was just sorted topics, which/whose topics?
that would make sense if we are using collections that form a subset of all topics somewhere in the code. otherwise it seems redundant

i'm not sure i follow. even if the topics weren't sorted, we would still ensure unassigned partitions are present, no? what makes them distribute more evenly?

Reply:
Distribute evenly topic wise as well, as in if I distributed partitions in random order its possible that all the partitions from a single topic go to one member only, if I rr a list that goes topic wise I'm distributing each topics partitions evenly

Reply:
thanks, that makes sense